### ISSUE TITLE
Undefined reference to `__enzyme_autodiff` with Clang 16

### ISSUE THREAD

**Issue by roastduck** [EnzymeAD/Enzyme#1227]
Hi team,

I'm compiling the following example program from the guide:

```c
#include <stdio.h>
extern double __enzyme_autodiff(void*, double);
double square(double x) { return x * x; }
double dsquare(double x) {
    // This returns the derivative of square or 2 * x
    return __enzyme_autodiff((void*)square, x);
}
int main() {
    for (double i = 1; i < 5; i++)
        printf("square(%f)=%f, dsquare(%f)=%f", i, square(i), i, dsquare(i));
}
```

The command is: `/path/to/clang test.c -Xclang -load -Xclang /path/to/ClangEnzyme-16.so -O2 -o test`.

But I got the following error:

```
/usr/bin/ld: /tmp/test-7f7c29.o: in function `dsquare':
test.c:(.text+0x18): undefined reference to `__enzyme_autodiff'
/usr/bin/ld: /tmp/test-7f7c29.o: in function `main':
test.c:(.text+0x37): undefined reference to `__enzyme_autodiff'
/usr/bin/ld: test.c:(.text+0x69): undefined reference to `__enzyme_autodiff'
/usr/bin/ld: test.c:(.text+0x99): undefined reference to `__enzyme_autodiff'
/usr/bin/ld: test.c:(.text+0xc9): undefined reference to `__enzyme_autodiff'
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
```

Tested with Enzyme 0.0.66. Clang version information prints as follows:

```
clang version 16.0.0
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin
```

It works fine with Clang 12, but I am really looking forward to use Clang 16 to port a program that uses C++20.

Thanks in advance.

**Comment by tgymnich** [comment]
Clang 16 requires using the new pass manager.
Invoking the new pass manager looks something like this: `-fpass-plugin=/path/to/ClangEnzyme-16.so -Xclang -load -Xclang /path/to/ClangEnzyme-16.so

**Comment by tgymnich** [comment]
@jedbrown we are still using `-load` as a hack to enable Enzyme to parse command line arguments like `-ezyme-preopt=false` etc.

**Comment by roastduck** [comment]
After I prepended the `-fpass-plugin=...` argument, I only got the following error on the same program:

```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocess[65/9254], and associated run script.
Stack dump:
0.      Program arguments: /utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16 -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name test.c -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=none -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -debugger-tuning=gdb -fcoverage-compilation-dir=/adb/tmp/doc-example-1 -resource-dir /utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/lib/clang/16 -internal-isystem /utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/lib/clang/16/include -internal-isystem /usr/local/include -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/11/../../../../x86_64-linux-gnu/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -O2 -fdebug-compilation-dir=/adb/tmp/doc-example-1 -ferror-limit 19 -fgnuc-version=4.2.1 -fcolor-diagnostics -vectorize-loops -vectorize-slp -fpass-plugin=/utils/enzyme/Enzyme-0.0.66/enzyme/build/Enzyme/ClangEnzyme-16.so -load /utils/enzyme/Enzyme-0.0.66/enzyme/build/Enzyme/ClangEnzyme-16.so -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/test-27720b.o -x c test.c
1.      <eof> parser at end of file
2.      Optimizer
 #0 0x000055bec2703b68 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x3103b68)
 #1 0x000055bec2701aae llvm::sys::RunSignalHandlers() (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x3101aae)
 #2 0x000055bec27041ed SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f0698762520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x000055bec2075023 llvm::AttributeList::getRetAlignment() const (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x2a75023)
 #5 0x00007f0698083516 clearFunctionAttributes(llvm::Function*) (/utils/enzyme/Enzyme-0.0.66/enzyme/build/Enzyme/ClangEnzyme-16.so+0x842516)
 #6 0x00007f0698096869 EnzymeLogic::CreatePrimalAndGradient(ReverseCacheKey const&&, TypeAnalysis&, AugmentedReturn const*, bool) (/utils/enzyme/Enzyme-0.0.66/enzyme/build/Enzyme/ClangEnzyme-16.so+0x855869)
 #7 0x00007f06980160be (anonymous namespace)::EnzymeBase::HandleAutoDiff(llvm::Instruction*, unsigned int, llvm::Value*, llvm::Type*, llvm::SmallVectorImpl<llvm::Value*>&, std::map<int, llvm::Type*, std::less<int>, std::allocator<std::pair<int const, llvm::Type*>>> const&, std::vector<DIFFE_TYPE, std::allocator<DIFFE_TYPE>> const&, llvm::Function*, DerivativeMode, (anonymous namespace)::EnzymeBase::Options&, bool) Enzyme.cpp:0:0
 #8 0x00007f06980180f9 (anonymous namespace)::EnzymeBase::HandleAutoDiffArguments(llvm::CallInst*, DerivativeMode, bool) Enzyme.cpp:0:0
 #9 0x00007f069801c436 (anonymous namespace)::EnzymeBase::lowerEnzymeCalls(llvm::Function&, std::set<llvm::Function*, std::less<llvm::Function*>, std::allocator<llvm::Function*>>&) Enzyme.cpp:0:0
#10 0x00007f069801d605 (anonymous namespace)::EnzymeBase::run(llvm::Module&) Enzyme.cpp:0:0
#11 0x00007f0698024f84 EnzymeNewPM::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/utils/enzyme/Enzyme-0.0[28/9254]e/build/Enzyme/ClangEnzyme-16.so+0x7e3f84)
#12 0x00007f069806ef57 llvm::detail::PassModel<llvm::Module, EnzymeNewPM, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/utils/enzyme/Enzyme-0.0.66/enzyme/build/Enzyme/ClangEnzyme-16.so+0x82df57)                                                                                                               #13 0x000055bec217b8a7 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x2b7b8a7)
#14 0x000055bec29bafb6 (anonymous namespace)::EmitAssemblyHelper::RunOptimizationPipeline(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>&, std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>&) BackendUtil.cpp:0:0                                                                              #15 0x000055bec29b18b6 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x33b18b6)
#16 0x000055bec35e45b4 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x3fe45b4)
#17 0x000055bec307d9dc clang::MultiplexConsumer::HandleTranslationUnit(clang::ASTContext&) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x3a7d9dc)
#18 0x000055bec441978d clang::ParseAST(clang::Sema&, bool, bool) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x4e1978d)
#19 0x000055bec3042fd7 clang::FrontendAction::Execute() (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x3a42fd7)
#20 0x000055bec2fbb354 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x39bb354)
#21 0x000055bec30fa8ef clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x3afa8ef)                                                                                       #22 0x000055bec08ed9e4 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x12ed9e4)                                                                                        #23 0x000055bec08ea656 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#24 0x000055bec08ea2f9 clang_main(int, char**) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x12ea2f9)
#25 0x00007f0698749d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#26 0x00007f0698749e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#27 0x000055bec08e659a _start (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x12e659a)
clang-16: error: unable to execute command: Segmentation fault (core dumped)
clang-16: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 16.0.0
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin
clang-16: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-16: note: diagnostic msg: /tmp/test-fdd409.c
clang-16: note: diagnostic msg: /tmp/test-fdd409.sh
clang-16: note: diagnostic msg:

********************
```

(I guess you don't need the displayed `/tmp/test-fdd409.c` and `/tmp/test-fdd409.sh`. Tell me if you do.)

**Comment by wsmoses** [comment]
@tgymnich I tried to reproduce on compiler explorer but it seemed to have a weird issue, can you check?

**Comment by tgymnich** [comment]
@wsmoses I cannot reproduce. Mind sending a link?
https://fwd.gymni.ch/th0DJG

**Comment by wsmoses** [comment]
well right now, cant see it: 
<img width="1112" alt="Screen Shot 2023-06-07 at 5 24 55 PM" src="https://github.com/EnzymeAD/Enzyme/assets/1260124/4cc61175-5e5a-472c-a975-f458fdb180c0">

**Comment by roastduck** [comment]
I updated to Enzyme 0.0.68 in my environment, but it resulted in another error:

```
clang (LLVM option parsing): CommandLine Error: Option 'enzyme-print-activity' registered more than once!
fatal error: error in backend: inconsistency in registered CommandLine options
clang-16: error: clang frontend command failed with exit code 70 (use -v to see invocation)
clang version 16.0.0
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin
clang-16: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-16: note: diagnostic msg: /tmp/test-0d9c38.c
clang-16: note: diagnostic msg: /tmp/test-0d9c38.sh
clang-16: note: diagnostic msg:
********************
```

**Comment by roastduck** [comment]
> @roastduck can you paste the exact full command line args used?


```
/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang test.c -fpass-plugin=/utils/enzyme/Enzyme-0.0.66/enzyme/build/Enzyme/ClangEnzyme-16.so -Xclang -load -Xclang /utils/enzyme/Enzyme-0.0.68/enzyme/build/Enzyme/ClangEnzyme-16.so -O2 -o test
```

The LLVM is from https://github.com/llvm/llvm-project/releases/download/llvmorg-16.0.0/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz.

Enzyme is 0.0.68 release, configured with `cmake .. -DLLVM_DIR=/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/lib/cmake/llvm -DCMAKE_MODULE_PATH=/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/lib/cmake/llvm`.

**Comment by wsmoses** [comment]
I'm not quite able to reproduce (using same Enzyme , but a debug clang):

```
wmoses@beast:~/git/Enzyme/enzyme/build16 ((HEAD detached at v0.0.68)) $ /home/wmoses/llvms/llvm16/buildD/./bin/clang test.c -fpass-plugin=Enzyme/ClangEnzyme-16.so -Xclang -load -Xclang Enzyme/ClangEnzyme-16.so -O2 -o test.exe
wmoses@beast:~/git/Enzyme/enzyme/build16 ((HEAD detached at v0.0.68)) $ ./test.exe 
square(1.000000)=1.000000, dsquare(1.000000)=2.000000square(2.000000)=4.000000, dsquare(2.000000)=4.000000square(3.000000)=9.000000, dsquare(3.000000)=6.000000square(4.000000)=16.000000, dsquare(4.000000)=8.000000wmoses@beast:~/git/Enzyme/enzyme/build16 ((HEAD detached at v0.0.68)) $ 
```

Are you certain that you are using the version of Enzyme you intend?

The -fpass-plugin says `/utils/enzyme/Enzyme-0.0.66/enzyme/build/Enzyme/ClangEnzyme-16.so` whereas your other loader says `/utils/enzyme/Enzyme-0.0.68/enzyme/build/Enzyme/ClangEnzyme-16.so`

**Comment by roastduck** [comment]
Sorry, it was my mistake. After I corrected it to `/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang test.c -fpass-plugin=/utils/enzyme/Enzyme-0.0.68/enzyme/build/Enzyme/ClangEnzyme-16.so -Xclang -load -Xclang /utils/enzyme/Enzyme-0.0.68/enzyme/build/Enzyme/ClangEnzyme-16.so -O2 -o test`, the original bug posted in this thread occured:

```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16 -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name test.c -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=none -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -debugger-tuning=gdb -fcoverage-compilation-dir=/adb/tmp/doc-example-1 -resource-dir /utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/lib/clang/16 -internal-isystem /utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/lib/clang/16/include -internal-isystem /usr/local/include -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/11/../../../../x86_64-linux-gnu/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -O2 -fdebug-compilation-dir=/adb/tmp/doc-example-1 -ferror-limit 19 -fgnuc-version=4.2.1 -fcolor-diagnostics -vectorize-loops -vectorize-slp -fpass-plugin=/utils/enzyme/Enzyme-0.0.68/enzyme/build/Enzyme/ClangEnzyme-16.so -load /utils/enzyme/Enzyme-0.0.68/enzyme/build/Enzyme/ClangEnzyme-16.so -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/test-15f8de.o -x c test.c
1.      <eof> parser at end of file
2.      Optimizer
 #0 0x000055c441503b68 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x3103b68)
 #1 0x000055c441501aae llvm::sys::RunSignalHandlers() (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x3101aae)
 #2 0x000055c4415041ed SignalHandler(int) Signals.cpp:0:0
 #3 0x00007fa170326520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x000055c440e75023 llvm::AttributeList::getRetAlignment() const (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x2a75023)
 #5 0x00007fa16fc2cebc clearFunctionAttributes(llvm::Function*) (/utils/enzyme/Enzyme-0.0.68/enzyme/build/Enzyme/ClangEnzyme-16.so+0x850ebc)
 #6 0x00007fa16fc406c2 EnzymeLogic::CreatePrimalAndGradient(ReverseCacheKey const&&, TypeAnalysis&, AugmentedReturn const*, bool) (/utils/enzyme/Enzyme-0.0.68/enzyme/build/Enzyme/ClangEnzyme-16.so+0x8646c2)
 #7 0x00007fa16fbbf828 (anonymous namespace)::EnzymeBase::HandleAutoDiff(llvm::Instruction*, unsigned int, llvm::Value*, llvm::Type*, llvm::SmallVectorImpl<llvm::Value*>&, std::map<int, llvm::Type*, std::less<int>, std::allocator<std::pair<int const, llvm::Type*>>> const&, std::vector<DIFFE_TYPE, std::allocator<DIFFE_TYPE>> const&, llvm::Function*, DerivativeMode, (anonymous namespace)::EnzymeBase::Options&, bool) Enzyme.cpp:0:0
 #8 0x00007fa16fbc1863 (anonymous namespace)::EnzymeBase::HandleAutoDiffArguments(llvm::CallInst*, DerivativeMode, bool) Enzyme.cpp:0:0
 #9 0x00007fa16fbc5ba0 (anonymous namespace)::EnzymeBase::lowerEnzymeCalls(llvm::Function&, std::set<llvm::Function*, std::less<llvm::Function*>, std::allocator<llvm::Function*>>&) Enzyme.cpp:0:0
#10 0x00007fa16fbc6d6f (anonymous namespace)::EnzymeBase::run(llvm::Module&) Enzyme.cpp:0:0
#11 0x00007fa16fbce730 EnzymeNewPM::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/utils/enzyme/Enzyme-0.0.68/enzyme/build/Enzyme/ClangEnzyme-16.so+0x7f2730)
#12 0x00007fa16fc18703 llvm::detail::PassModel<llvm::Module, EnzymeNewPM, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/utils/enzyme/Enzyme-0.0.68/enzyme/build/Enzyme/ClangEnzyme-16.so+0x83c703)
#13 0x000055c440f7b8a7 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x2b7b8a7)
#14 0x000055c4417bafb6 (anonymous namespace)::EmitAssemblyHelper::RunOptimizationPipeline(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>&, std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>&) BackendUtil.cpp:0:0
#15 0x000055c4417b18b6 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x33b18b6)
#16 0x000055c4423e45b4 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x3fe45b4)
#17 0x000055c441e7d9dc clang::MultiplexConsumer::HandleTranslationUnit(clang::ASTContext&) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x3a7d9dc)
#18 0x000055c44321978d clang::ParseAST(clang::Sema&, bool, bool) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x4e1978d)
#19 0x000055c441e42fd7 clang::FrontendAction::Execute() (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x3a42fd7)
#20 0x000055c441dbb354 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x39bb354)
#21 0x000055c441efa8ef clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x3afa8ef)
#22 0x000055c43f6ed9e4 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x12ed9e4)
#23 0x000055c43f6ea656 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#24 0x000055c43f6ea2f9 clang_main(int, char**) (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x12ea2f9)
#25 0x00007fa17030dd90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#26 0x00007fa17030de40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#27 0x000055c43f6e659a _start (/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang-16+0x12e659a)
clang-16: error: unable to execute command: Segmentation fault (core dumped)
clang-16: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 16.0.0
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin
clang-16: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-16: note: diagnostic msg: /tmp/test-8af71d.c
clang-16: note: diagnostic msg: /tmp/test-8af71d.sh
clang-16: note: diagnostic msg:

********************
```

To reproduce, I have made a Dockerfile that reflects my environment:

```
FROM nvidia/cuda:11.8.0-devel-ubuntu22.04

RUN apt-get update
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
        wget \
        vim \
        git \
        cmake \
        libtinfo5 libtinfo-dev \
        autoconf automake libtool \
        && rm -rf /var/lib/apt/lists/*

# Install Clang 16
WORKDIR /utils/clang
RUN wget https://github.com/llvm/llvm-project/releases/download/llvmorg-16.0.0/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz \
    && tar -xf clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz
ENV PATH=/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/:$PATH
ENV LD_LIBRARY_PATH=/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/lib:$LD_LIBRARY_PATH

# Install Enzyme
WORKDIR /utils/enzyme
RUN wget https://github.com/EnzymeAD/Enzyme/archive/refs/tags/v0.0.69.tar.gz \
    && tar -xf v0.0.69.tar.gz
WORKDIR /utils/enzyme/Enzyme-0.0.69/enzyme/build
RUN cmake .. -DLLVM_DIR=/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/lib/cmake/llvm -DCMAKE_MODULE_PATH=/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/lib/cmake/llvm \
    && make -j

WORKDIR /workspace

ENTRYPOINT ["/bin/bash"]
```

To reproduce:

1. Save it in `reproducer.Dockerfile`.
2. Run `docker build -f reproducer.Dockerfile -t "enzyme-bug-reproducer:v0" .` to build.
3. Enter the image by `docker run -it enzyme-bug-reproducer:v0`.
4. Paste the program in `test.c`.
5. Run `/utils/clang/clang+llvm-16.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang test.c -fpass-plugin=/utils/enzyme/Enzyme-0.0.69/enzyme/build/Enzyme/ClangEnzyme-16.so -Xclang -load -Xclang /utils/enzyme/Enzyme-0.0.69/enzyme/build/Enzyme/ClangEnzyme-16.so -O2 -o test` (I have updated to Enzyme 0.0.69)

You will see the error message above.

**Comment by wsmoses** [comment]
@roastduck the way plugins work has now changed to just simply adding `-fplugin=/path/to/ClangEnzyme-XX.so`, mind checking if it still errors?
